src/vidpf.rs

```diff
@@ -0,0 +1,667 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Verifiable Incremental Distributed Point Function (VIDPF).
+//!
+//! The VIDPF construction is specified in [[draft-mouris-cfrg-mastic]] and builds
+//! on techniques from [[MST23]] and [[CP22]] to lift an IDPF to a VIDPF.
+//!
+//! [CP22]: https://eprint.iacr.org/2021/580
+//! [MST23]: https://eprint.iacr.org/2023/080
+//! [draft-mouris-cfrg-mastic]: https://datatracker.ietf.org/doc/draft-mouris-cfrg-mastic/01/
+
+use core::{
+    borrow::Borrow,
+    iter::zip,
+    marker::PhantomData,
+    ops::{Add, BitAnd, BitXor, Sub},
+};
+
+use rand_core::RngCore;
+use subtle::{Choice, ConditionallyNegatable, ConditionallySelectable};
+
+use crate::{
+    field::FieldElement,
+    vdaf::{
+        xof::{Seed as XofSeed, Xof, XofFixedKeyAes128, XofTurboShake128},
+        VdafError,
+    },
+};
+
+/// Creates a VIDPF instance where the function F: {0.1}* --> [Range].
+#[must_use]
+pub const fn new<R>(bind_info: Vec<u8>) -> Instance<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    Instance {
+        bind_info,
+        _cd: PhantomData,
+    }
+}
+
+/// Contains the global parameters of a VIDPF instance.
+///
+/// The function is defined as F: {0.1}* --> [Range].
+pub struct Instance<R>
+where
+    // Represents the codomain of a Distributed Point Function (DPF).
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    // Used to cryptographically bind some information.
+    bind_info: Vec<u8>,
+
+    _cd: PhantomData<R>,
+}
+
+impl<R> Instance<R>
+where
+    R: Range,
+    for<'a> &'a R: RangeOps<R>,
+{
+    /// The gen method splits the point function `F(alpha) = beta` in two private keys
+    /// used by the aggregation servers, and a common public key.
+    /// The `beta` value must be a non-zero value.
+    ///
+    /// # Errors
+    /// Triggers an error if the random generator fails, or when beta is zero.
+    pub fn gen(&self, alpha: &[u8], beta: &R) -> Result<(PublicKey<R>, Key, Key), VdafError> {
+        if beta.is_zero() {
+            return Err(VidpfError::InvalidParameter("beta cannot be zero".into()).into());
+        }
+
+        // Key Generation.
+        let k0 = Key::gen(ServerID::S0)?;
+        let k1 = Key::gen(ServerID::S1)?;
+
+        let mut s_i = [Seed::from(&k0), Seed::from(&k1)];
+        let mut t_i = [ControlBit::from(&k0), ControlBit::from(&k1)];
+
+        let n = 8 * alpha.len();
+        let mut cw = Vec::with_capacity(n);
+        let mut cs = Vec::with_capacity(n);
+
+        for i in 0..n {
+            let alpha_i = ControlBit::from((alpha[i / 8] >> (i % 8)) & 0x1);
```

Comment:
Question: would bit operations be faster than `/` and `%`, or is the compiler smart enough to recognize it can do bit fiddling here?

(**Don't just blindly take the suggestion below, I'm not 100% I got the bit fiddling right!**)
```suggestion
            let alpha_i = ControlBit::from((alpha[i >> 3] >> (i & 0xff)) & 0x1);
```

Reply:
no need, as it's using bitvector now.